File: dotnet/Net/Multiplayer/Multiplayer.cs
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

using System;
using System.IO;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using System.Net.Security;
using System.Net.NetworkInformation;
using System.Text;
using System.Runtime;
using System.Runtime.Serialization;


namespace Michitai.Lan.Net.Multiplayer
{
    /// <summary>
    /// Static class providing centralized multiplayer game management and configuration.
    /// </summary>
    public static class Multiplayer
    {
        /// <summary>
        /// The name of the multiplayer game.
        /// </summary>
        public static string Name = "New Multiplayer Game";

        /// <summary>
        /// The IP address to which the server or/and client will connect.
        /// </summary>
        public static IPAddress IpAddress = IPAddress.Any;

        /// <summary>
        /// The port to which the client will connect.
        /// </summary>
        public static int Port = 50000;

        /// <summary>
        /// Port range. One of them will be used by the server. Default 50000-50128.
        /// </summary>
        public static PortRange ServerPortRange = new PortRange(50000, 50128);

        /// <summary>
        /// Port range. Used by Server and Client to Response and Request UDP Messages. Default 60000-60128.
        /// </summary>
        public static PortRange BroadcastPortRange = new PortRange(60000, 60128);

        /// <summary>
        /// The multiplayer server instance.
        /// </summary>
        public static Server Server = null;

        /// <summary>
        /// The multiplayer client instance.
        /// </summary>
        public static Client Client = null;

        /// <summary>
        /// The broadcast server instance.
        /// </summary>
        public static BroadcastServer BroadcastServer = null;

        /// <summary>
        /// Task for running the broadcast server.
        /// </summary>
        private static Task _BroadcastServerTask = null;

        /// <summary>
        /// Cancellation token source for the broadcast server task.
        /// </summary>
        private static CancellationTokenSource _BroadcastServerTaskTokenSource = null;

        /// <summary>
        /// The broadcast client instance.
        /// </summary>
        public static BroadcastClient BroadcastClient = null;

        /// <summary>
        /// Gets whether the broadcast client is locating responses from the server.
        /// </summary>
        /// <returns>True if BroadcastClient locating Response from Server, otherwise False.</returns>
        public static bool BroadcastClientLocating
    {
        get
        {
            lock (_BroadcastClientLocatingLock)
            {
                return _BroadcastClientLocating;
            }
        }

        private set
        {
            lock (_BroadcastClientLocatingLock)
            {
                _BroadcastClientLocating = value;
            }
        }
    }

    /// <summary>
    /// 
    /// </summary>
    private static bool _BroadcastClientLocating = false;

    /// <summary>
    /// 
    /// </summary>
    private static readonly object _BroadcastClientLocatingLock = new object();

    /// <summary>
    /// 
    /// </summary>
    private static Task _BroadcastClientTask = null;

    /// <summary>
    /// 
    /// </summary>
    private static CancellationTokenSource _BroadcastClientTaskTokenSource = null;


    /// <summary>
    /// 
    /// </summary>
    public static Queue<Command> ClientCommands = new Queue<Command>();

    /// <summary>
    /// Number of requests that can be sent at the same time.
    /// </summary>
    public static int ClientOnceMaxCommands = 4;



    /// <summary>
    /// 
    /// </summary>
    public static bool IsServer => Server != null;

    /// <summary>
    /// 
    /// </summary>
    public static bool IsClient => Client != null;



    /// <summary>
    /// 
    /// </summary>
    public static event Action OnStartServer;

    /// <summary>
    /// 
    /// </summary>
    public static event Action OnStartClient;

    /// <summary>
    /// 
    /// </summary>
    public static event Action OnClientStarted;

    /// <summary>
    /// 
    /// </summary>
    public static event Action OnServerStarted;



    /// <summary>
    /// 
    /// </summary>
    /// <param name="serverGameData"></param>
    public static void StartServer(EPlatform platform, ServerGameData serverGameData, BroadcastServer.ProcessMessageDelegate processMessage, int receiveRequestsDelayMilliseconds = 100)
    {
        if (IsServer)
            return;


        Server = new Server(Name, serverGameData, IpAddress, ServerPortRange);

        IpAddress = Server.IPEndPoint.Address;

        Port = Server.IPEndPoint.Port;


        _BroadcastServerTaskTokenSource = new CancellationTokenSource();

        var token = _BroadcastServerTaskTokenSource.Token;


        _BroadcastServerTask = Task.Run(async () =>
        {
            token.ThrowIfCancellationRequested();

            BroadcastServer = new BroadcastServer(IpAddress, BroadcastPortRange);

            while (true)
            {
                token.ThrowIfCancellationRequested();

                if ((platform & EPlatform.Android) == EPlatform.Android)
                {
                    BroadcastServer.Broadcast(processMessage);
                }
                else if((platform & EPlatform.Windows) == EPlatform.Windows)
                {
                    await BroadcastServer.BroadcastAsync(processMessage);
                }
                else
                {
                    throw new NotImplementedException($"Platform: {platform}");
                }

                token.ThrowIfCancellationRequested();

                await Task.Delay(receiveRequestsDelayMilliseconds, _BroadcastServerTaskTokenSource.Token);
            }

        }, _BroadcastServerTaskTokenSource.Token);


        OnStartServer?.Invoke();

        Server.Start();

        OnServerStarted?.Invoke();
    }

    /// <summary>
    /// 
    /// </summary>
    public static void StartClient()
    {
        if (IsClient)
            return;


        Client = new Client(IpAddress, Port);


        ClientCommands = new Queue<Command>();


        OnStartClient?.Invoke();

        Client.Start();

        OnClientStarted?.Invoke();
    }


    /// <summary>
    /// 
    /// </summary>
    public static void StopServer()
    {
        DebugConsole.LogWarning("[MULTIPLAYER] Cancelling BroadcastServerTask...");

        _BroadcastServerTaskTokenSource?.Cancel();

        _BroadcastClientTaskTokenSource = null;

        _BroadcastServerTask = null;

        DebugConsole.Log("[MULTIPLAYER] BroadcastServerTask canceled.");


        DebugConsole.LogWarning("[MULTIPLAYER] Stopping Server...");

        Server?.Stop();

        Server = null;

        DebugConsole.Log("[MULTIPLAYER] Server stopped.");


        DebugConsole.LogWarning("[MULTIPLAYER] Stopping BroadcastServer...");

        BroadcastServer?.Stop();

        BroadcastServer = null;

        DebugConsole.Log("[MULTIPLAYER] BroadcastServer stopped.");
    }

    /// <summary>
    /// 
    /// </summary>
    public static void StopClient()
    {
        Client?.Stop();

        Client = null;
    }

    /// <summary>
    /// 
    /// </summary>
    public static void Stop()
    {
        StopClient();

        StopServer();
    }

    /// <summary>
    /// 
    /// </summary>
    public static void ClearEvents()
    {
        OnStartServer = null;

        OnServerStarted = null;

        OnStartClient = null;

        OnClientStarted = null;
    }



    /// <summary>
    /// 
    /// </summary>
    public static void StartBroadcastClient(EPlatform platform, AppMessage request, BroadcastClient.OnReceiveResponseDelegate onReceiveResponse, int receiveResponsesMilliseconds = 5000, int repeatAfterMilliseconds = 5000)
    {
        if (_BroadcastClientTask != null)
            return;

        _BroadcastClientTaskTokenSource = new CancellationTokenSource();

        var token = _BroadcastClientTaskTokenSource.Token;

        if ((platform & EPlatform.Windows) == EPlatform.Windows || (platform & EPlatform.Linux) == EPlatform.Linux || (platform & EPlatform.MacOS) == EPlatform.MacOS)
        {
            _BroadcastClientTask = Task.Run(async () =>
            {
                token.ThrowIfCancellationRequested();

                BroadcastClient = new BroadcastClient(IpAddress, BroadcastPortRange);

                while (true)
                {
                    token.ThrowIfCancellationRequested();

                    BroadcastClientLocating = true;

                    IPAddress[] masks = null;

                    bool success = Lan.TryGetLocalIPv4Masks(platform, out masks);

                    if (success)
                    {
                        await BroadcastClient.BroadcastRequestAsync(masks, BroadcastPortRange, request, onReceiveResponse, receiveResponsesMilliseconds);
                    }
                    else
                    {
                        await BroadcastClient.BroadcastRequestAsync(IPAddress.Broadcast, BroadcastPortRange, request, onReceiveResponse, receiveResponsesMilliseconds);
                    }

                    BroadcastClientLocating = false;

                    token.ThrowIfCancellationRequested();

                    await Task.Delay(repeatAfterMilliseconds, _BroadcastClientTaskTokenSource.Token);
                }

            }, _BroadcastClientTaskTokenSource.Token);
        }
        else if((platform & EPlatform.Android) == EPlatform.Android || (platform & EPlatform.IOS) == EPlatform.IOS)
        {
            Task.Run(() =>
            {
                token.ThrowIfCancellationRequested();

                BroadcastClient = new BroadcastClient(IpAddress, BroadcastPortRange);

                while (true)
                {
                    token.ThrowIfCancellationRequested();

                    BroadcastClientLocating = true;

                    IPAddress[] masks = null;

                    bool success = Lan.TryGetLocalIPv4Masks(platform, out masks);

                    if (success)
                    {
                        BroadcastClient.BeginBroadcastRequest(masks, BroadcastPortRange, request, onReceiveResponse, receiveResponsesMilliseconds);
                    }
                    else
                    {
                        BroadcastClient.BeginBroadcastRequest(IPAddress.Broadcast, BroadcastPortRange, request, onReceiveResponse, receiveResponsesMilliseconds);
                    }

                    BroadcastClientLocating = false;

                    token.ThrowIfCancellationRequested();

                    Task.Delay(repeatAfterMilliseconds, _BroadcastClientTaskTokenSource.Token).Wait();
                }

            }, _BroadcastClientTaskTokenSource.Token);
        }
        else
        {
            throw new NotImplementedException();
        }
    }

    /// <summary>
    /// 
    /// </summary>
    public static void StopBroadcastClient()
    {
        _BroadcastClientTaskTokenSource?.Cancel();

        _BroadcastClientTaskTokenSource = null;

        _BroadcastClientTask = null;


        BroadcastClient?.Stop();

        BroadcastClient = null;


        BroadcastClientLocating = false;
    }
}
}

File: cpp/src/Net/Multiplayer/Multiplayer.cpp
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

#include <Michitai/Lan/Net/Multiplayer/Multiplayer.hpp>
#include <Michitai/Lan/Debugging/DebugConsole.hpp>

#include <chrono>


namespace Michitai::Lan::Net::Multiplayer
{
    // Type aliases — inside Multiplayer member functions the member names
    // (Server, Client, BroadcastServer, BroadcastClient) shadow the types.
    namespace
    {
        using ServerT          = Michitai::Lan::Net::Multiplayer::Server;
        using ClientT          = Michitai::Lan::Net::Multiplayer::Client;
        using BroadcastServerT = Michitai::Lan::Net::Multiplayer::BroadcastServer;
        using BroadcastClientT = Michitai::Lan::Net::Multiplayer::BroadcastClient;
    }

    void Multiplayer::SleepInterruptible(int milliseconds, const std::stop_token& token)
    {
        const auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(milliseconds);

        while (!token.stop_requested())
        {
            const auto remaining = deadline - std::chrono::steady_clock::now();
            if (remaining <= std::chrono::milliseconds(0))
                break;

            std::this_thread::sleep_for(
                std::min<std::chrono::steady_clock::duration>(remaining, std::chrono::milliseconds(10)));
        }
    }

    void Multiplayer::StartServer(EPlatform platform,
                                  const Data::ServerGameDataPtr& serverGameData,
                                  const ProcessMessageDelegate& processMessage,
                                  int receiveRequestsDelayMilliseconds)
    {
        (void)platform; // All platforms share the same socket implementation in the C++ port.

        if (IsServer())
            return;

        Server = ServerT::Create(Name, serverGameData, IpAddress, ServerPortRange);

        IpAddress = Server->IpEndPoint().Address;
        Port = Server->IpEndPoint().Port;

        BroadcastServer = std::make_shared<BroadcastServerT>(IpAddress, BroadcastPortRange);

        _broadcastServerStop = std::stop_source();
        const std::stop_token token = _broadcastServerStop.get_token();
        const BroadcastServerT::Ptr broadcastServer = BroadcastServer;

        _broadcastServerThread = std::thread(
            [token, processMessage, broadcastServer, receiveRequestsDelayMilliseconds]
        {
            while (!token.stop_requested())
            {
                // One receive+respond cycle per iteration; the 100 ms poll keeps
                // the loop responsive to Stop().
                broadcastServer->Broadcast(processMessage, 100);

                SleepInterruptible(receiveRequestsDelayMilliseconds, token);
            }
        });

        Server->OnState += [](const IPEndPoint& point, const Message& message)
        {
            OnState.Invoke(point, message);
        };

        OnStartServer.Invoke();
        Server->Start();
        OnServerStarted.Invoke();
    }

    void Multiplayer::StartClient()
    {
        if (IsClient())
            return;

        Client = ClientT::Create(IpAddress, Port);

        ClientCommands = std::queue<Commands::Command::Ptr>();

        Client->OnState += [](const IPEndPoint& point, const Message& message)
        {
            OnState.Invoke(point, message);
        };

        OnStartClient.Invoke();
        Client->Start();
        OnClientStarted.Invoke();
    }

    void Multiplayer::StopServer()
    {
        Debug::DebugConsole::LogWarning("[MULTIPLAYER] Cancelling BroadcastServerTask...");

        _broadcastServerStop.request_stop();
        if (_broadcastServerThread.joinable())
            _broadcastServerThread.join();

        Debug::DebugConsole::Log("[MULTIPLAYER] BroadcastServerTask canceled.");

        Debug::DebugConsole::LogWarning("[MULTIPLAYER] Stopping Server...");

        if (Server)
            Server->Stop();
        Server = nullptr;

        Debug::DebugConsole::Log("[MULTIPLAYER] Server stopped.");

        Debug::DebugConsole::LogWarning("[MULTIPLAYER] Stopping BroadcastServer...");

        if (BroadcastServer)
            BroadcastServer->Stop();
        BroadcastServer = nullptr;

        Debug::DebugConsole::Log("[MULTIPLAYER] BroadcastServer stopped.");
    }

    void Multiplayer::StopClient()
    {
        if (Client)
            Client->Stop();
        Client = nullptr;
    }

    void Multiplayer::Stop()
    {
        StopClient();
        StopServer();
    }

    void Multiplayer::StartBroadcastClient(EPlatform platform,
                                           const AppMessage& request,
                                           const OnReceiveResponseDelegate& onReceiveResponse,
                                           int receiveResponsesMilliseconds,
                                           int repeatAfterMilliseconds)
    {
        if (_broadcastClientThread.joinable())
            return;

        _broadcastClientStop = std::stop_source();
        const std::stop_token token = _broadcastClientStop.get_token();

        BroadcastClient = std::make_shared<BroadcastClientT>(IpAddress, BroadcastPortRange);
        const BroadcastClientT::Ptr broadcastClient = BroadcastClient;

        _broadcastClientThread = std::thread(
            [token, broadcastClient, platform, request, onReceiveResponse,
             receiveResponsesMilliseconds, repeatAfterMilliseconds]
        {
            while (!token.stop_requested())
            {
                SetBroadcastClientLocating(true);

                std::vector<IPAddress> masks;
                const bool success = Lan::TryGetLocalIPv4Masks(platform, masks);

                // Send the discovery probe to every subnet mask x every broadcast port.
                if (success)
                {
                    for (const IPAddress& mask : masks)
                    {
                        for (int port = BroadcastPortRange.First; port <= BroadcastPortRange.Last; port++)
                        {
                            try { broadcastClient->Socket().Send(mask, port, request); }
                            catch (...) {}
                        }
                    }
                }
                else
                {
                    for (int port = BroadcastPortRange.First; port <= BroadcastPortRange.Last; port++)
                    {
                        try { broadcastClient->Socket().Send(IPAddress::Broadcast(), port, request); }
                        catch (...) {}
                    }
                }

                // Collect responses for the configured window.
                broadcastClient->CollectResponses(onReceiveResponse, receiveResponsesMilliseconds, token);

                SetBroadcastClientLocating(false);

                if (token.stop_requested())
                    break;

                SleepInterruptible(repeatAfterMilliseconds, token);
            }
        });
    }

    void Multiplayer::StopBroadcastClient()
    {
        _broadcastClientStop.request_stop();
        if (_broadcastClientThread.joinable())
            _broadcastClientThread.join();

        if (BroadcastClient)
            BroadcastClient->Stop();
        BroadcastClient = nullptr;

        SetBroadcastClientLocating(false);
    }
}
